server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java

```diff
@@ -0,0 +1,265 @@
+
+        if (existingKudos.getDateApproved() != null) {
+            throw new BadArgException("Kudos with id %s has already been approved".formatted(kudosId));
+        }
+
+        existingKudos.setDateApproved(LocalDate.now());
+
+        return kudosRepository.update(existingKudos);
+    }
+
+    @Override
+    public KudosResponseDTO getById(UUID id) {
+
+        Kudos kudos = kudosRepository.findById(id).orElseThrow(() ->
+                new NotFoundException(KUDOS_DOES_NOT_EXIST_MSG.formatted(id)));
+
+        UUID currentUserId = currentUserServices.getCurrentUser().getId();
+        boolean isSender = currentUserId.equals(kudos.getSenderId());
+
+        if (kudos.getDateApproved() == null) {
+            // If not yet approved, only admins and the sender can access the kudos
+            if (!currentUserServices.isAdmin() && !isSender) {
+                throw new PermissionException(NOT_AUTHORIZED_MSG);
+            }
+        } else {
+            // If approved, admins, the sender, and the recipient can access the kudos
+
+            List<KudosRecipient> recipients = kudosRecipientRepository.findByKudosId(id);
+            boolean isRecipient = recipients
+                    .stream()
+                    .anyMatch(recipient -> recipient.getMemberId().equals(currentUserId));
+
+            if (!currentUserServices.isAdmin() && !isSender && !isRecipient) {
+                throw new PermissionException(NOT_AUTHORIZED_MSG);
+            }
+        }
+
+        return constructKudosResponseDTO(kudos);
+
+    }
+
+    @Override
+    public void delete(UUID id) {
+        if (!currentUserServices.isAdmin()) {
+            throw new PermissionException(NOT_AUTHORIZED_MSG);
+        }
+
+        Kudos kudos = kudosRepository.findById(id).orElseThrow(() ->
+                new NotFoundException(KUDOS_DOES_NOT_EXIST_MSG.formatted(id)));
+
+        // Delete all KudosRecipients associated with this kudos
+        List<KudosRecipient> recipients = kudosRecipientServices.getAllByKudosId(kudos.getId());
+        kudosRecipientRepository.deleteAll(recipients);
+
+        kudosRepository.deleteById(id);
+    }
+
+    @Override
+    public List<KudosResponseDTO> findByValues(@Nullable UUID recipientId, @Nullable UUID senderId, @Nullable Boolean isPending) {
+        if (isPending != null) {
+            return findByPending(Boolean.TRUE.equals(isPending));
+        } else if (recipientId != null) {
+            return findAllToMember(recipientId);
+        } else if (senderId != null) {
+            return findAllFromMember(senderId);
+        } else {
+            if (!currentUserServices.isAdmin()) {
+                throw new PermissionException(NOT_AUTHORIZED_MSG);
+            }
+
+            return kudosRepository.findAll()
+                    .stream()
+                    .map(this::constructKudosResponseDTO)
+                    .toList();
+        }
+    }
```

Comment:
Not 100% sure does what is expected... 

So if `isPending` is set to true or false, then `findByPending` is called, and `recipientId` and `senderId` are effectively ignored...

I'm not sure if there's a requirement for different behavior, but I just wanted to mention the point

Reply:
Yeah this is non-ideal. Not broken from a frontend perspective, but not ideal. I'll likely open an issue for this.